language/move-prover/src/prover_task_runner.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Prover task runner that runs multiple instances of the prover task and returns
+//! as soon as the fastest instance finishes.
+
+use crate::cli::Options;
+use async_std::task;
+use async_trait::async_trait;
+use futures::{future::FutureExt, pin_mut, select};
+use rand::Rng;
+use std::{
+    process::Output,
+    sync::mpsc::{channel, Sender},
+};
+use tokio::{
+    process::Command,
+    runtime::Runtime,
+    sync::{broadcast, broadcast::Receiver},
+};
+
+#[derive(Debug, Clone)]
+enum BroadcastMsg {
+    Stop,
+}
+
+#[async_trait]
+pub trait ProverTask {
+    type TaskResult: Send + 'static;
+    type TaskId: Send + Copy + 'static;
+
+    /// Initialize the task runner given the number of instances.
+    fn init(&mut self, num_instances: usize) -> Vec<Self::TaskId>;
+
+    /// Run the task with task_id. This function will be called from one of the worker threads.
+    async fn run(&mut self, task_id: Self::TaskId) -> Self::TaskResult;
+}
+
+pub struct ProverTaskRunner();
+
+impl ProverTaskRunner {
+    /// Run `num_instances` instances of the prover `task` and returns the task id
+    /// as well as the result of the fastest running instance.
+    pub fn run_tasks<T>(mut task: T, num_instances: usize) -> (T::TaskId, T::TaskResult)
+    where
+        T: ProverTask + Clone + Send + 'static,
+    {
+        let rt = Runtime::new().unwrap();
+
+        // Create channels for communication.
+        let (worker_tx, master_rx) = channel();
+        let (master_tx, _): (
+            tokio::sync::broadcast::Sender<BroadcastMsg>,
+            Receiver<BroadcastMsg>,
+        ) = broadcast::channel(num_instances);
+
+        // Initialize the prover tasks.
+        let task_ids = task.init(num_instances);
+        for task_id in task_ids {
+            let send_n = worker_tx.clone();
+            let worker_rx = master_tx.subscribe();
+            let cloned_task = task.clone();
+            // Spawn a task worker for each task_id.
+            rt.spawn(async move {
+                task::block_on(Self::run_task_until_cancelled(
```

Comment:
You probably want to call `.await` on this instead of using `task::block_on` from the `async-std` crate. In fact if thats the only use of the `async-std` crate you might want to drop it as a dependency.